src/Internal/Payment/State/PaymentState.php

```diff
@@ -0,0 +1,117 @@
+<?php
+/**
+ * Class PaymentState
+ *
+ * @package WooCommerce\Payments
+ */
+
+namespace WCPay\Internal\Payment\State;
+
+use WCPay\Internal\Payment\Exception\MethodUnavailableException;
+use WCPay\Internal\Payment\Exception\StateTransitionException;
+use WCPay\Internal\Payment\PaymentContext;
+
+/**
+ * Base class for payment states.
+ */
+abstract class PaymentState {
+	/**
+	 * Holds the context of the payments.
+	 *
+	 * The context contains working payment data,
+	 * while states are purely functional and stateless.
+	 *
+	 * @var PaymentContext
+	 */
+	private $context;
+
+	/**
+	 * A factory for states.
+	 *
+	 * States can
+	 *
+	 * @var StateFactory
+	 */
+	private $state_factory;
+
+	/**
+	 * Class constructor, only meant for storing dependencies.
+	 *
+	 * @param StateFactory $state_factory Factory for payment states.
+	 */
+	public function __construct( StateFactory $state_factory ) {
+		$this->state_factory = $state_factory;
+	}
+
+	/**
+	 * Stores the payment context.
+	 *
+	 * @param PaymentContext $context Payment context.
+	 */
+	public function set_context( PaymentContext $context ) {
+		$this->context = $context;
+	}
+
+	/**
+	 * Returns the payment context.
+	 *
+	 * @return PaymentContext
+	 */
+	public function get_context(): PaymentContext {
+		return $this->context;
+	}
+
+	/**
+	 * Creates a new instance of a given payment state class.
+	 *
+	 * States control the payment flow, and allow transitions to the next state.
+	 * This method should only be called whenever the process is ready to transition
+	 * to the next state, as each new state will be considered the payment's latest one.
+	 *
+	 * @param string $state_class The class of the state to crate.
+	 * @return PaymentState
+	 * @throws StateTransitionException In case the new state could not be created.
+	 */
+	protected function create_state( string $state_class ) {
```

Comment:
I suggest refactoring this function to something like `transition_to(string $state_class)` and retaining the `state_factory` within the context. However, feel free to disregard this comment if you prefer an alternative approach.

Reply:
I'd prefer the name `transition_to` if we were using the state pattern. With the current state machine, it's not 100% clear what a transition is:

- From within the states, transitions happen exactly in this method. We will probably also use it to store the payment and/or context in the database.
- However, externally, a transition method (ex. `process`) returns the next state, and one could argue that the transition happens exactly when the state is returned there. If you call `$state->process()` without storing the resulting reference (the one to the next state), the payment would essentially not have transitioned.

Reply:
Ok, that makes sense. Thank you for clarifying it. I think I mixed both concepts in this PR.